sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryParallelReadStreamSource.java

```diff
@@ -0,0 +1,201 @@
+    this.readLocation = checkNotNull(readLocation, "readLocation");
+    this.readSizeBytes = readSizeBytes;
+  }
+
+  @Override
+  public List<? extends BoundedSource<T>> split(
+      long desiredBundleSizeBytes, PipelineOptions options) {
+    // Stream source objects can't (currently) be split.
+    return ImmutableList.of(this);
+  }
+
+  @Override
+  public long getEstimatedSizeBytes(PipelineOptions pipelineOptions) {
+    return readSizeBytes != null ? readSizeBytes : 0;
+  }
+
+  @Override
+  public BoundedReader<T> createReader(PipelineOptions pipelineOptions) {
+    return new BigQueryParallelReadStreamReader<>(
+        readSession,
+        readLocation,
+        (readSessionOptions != null && readSessionOptions.getRowBatchSize() != null)
+            ? readSessionOptions.getRowBatchSize()
+            : 1000,
+        parseFn,
+        this,
+        bqServices,
+        pipelineOptions.as(BigQueryOptions.class));
+  }
+
+  @Override
+  public Coder<T> getOutputCoder() {
+    return coder;
+  }
+
+  /**
+   * Iterates over all rows assigned to a particular reader in a read session.
+   */
+  @Experimental(Experimental.Kind.SOURCE_SINK)
+  static class BigQueryParallelReadStreamReader<T> extends BoundedReader<T> {
+
+    private Session session;
+    private final BigQueryServices client;
+    private Iterator<ReadRowsResponse> responseIterator;
+    private Iterator<Row> rowIterator;
+    private final ReadRowsRequest request;
+    private SerializableFunction<SchemaAndRowProto, T> parseFn;
+    private BoundedSource<T> source;
+    private BigQueryOptions options;
+    private Row currentRow;
+
+    BigQueryParallelReadStreamReader(
+        Session session,
+        ReadLocation readLocation,
+        Integer rowBatchSize,
+        SerializableFunction<SchemaAndRowProto, T> parseFn,
+        BoundedSource<T> source,
+        BigQueryServices client,
+        BigQueryOptions options) {
+      this.session = checkNotNull(session, "session");
+      this.client = checkNotNull(client, "client");
+      this.parseFn = checkNotNull(parseFn, "parseFn");
+      this.source = checkNotNull(source, "source");
+      this.options = options;
+
+      this.request = ReadRowsRequest.newBuilder()
+          .setReadLocation(readLocation)
+          .setOptions(ReadOptions.newBuilder()
+              .setMaxRows(checkNotNull(rowBatchSize, "rowBatchSize")))
+          .build();
+    }
+
+    /**
+     * Empty operation, the table is already open for read.
+     * @throws IOException on failure
+     */
+    @Override
+    public boolean start() throws IOException {
+      responseIterator = client.getTableReadService(options).readRows(request);
+      return advance();
+    }
+
+    @Override
+    public boolean advance() {
+      if (rowIterator == null || !rowIterator.hasNext()) {
+        if (!responseIterator.hasNext()) {
+          currentRow = null;
+          return false;
+        }
```

Comment:
Looks like this does not add support for progress reporting and dynamic work rebalancing. Do you hope to add those features before merging ? Otherwise this can result in a significant performance regression compared to existing source (and Dataflow autoscaling will not work properly).

Reply:
Our API doesn't support splitting at the moment, so adding support for dynamic work rebalancing is not a priority in this release. We'll be adding support for stream splitting as part of our Dataflow alpha. Note this is also not new behavior as part of this change.